substrate/primitives/runtime-interface/test-wasm/src/lib.rs

```diff
@@ -56,76 +63,76 @@ pub trait TestApi {
 	/// We return a `Vec<u32>` because this will use the code path that uses SCALE
 	/// to pass the data between native/wasm. (`Vec<u8>` is passed without encoding the
 	/// data)
-	fn return_16kb() -> Vec<u32> {
+	fn return_16kb() -> AllocateAndReturnByCodec<Vec<u32>> {
 		vec![0; 4 * 1024]
 	}
 
-	fn return_option_vec() -> Option<Vec<u8>> {
+	fn return_option_vec() -> AllocateAndReturnByCodec<Option<Vec<u8>>> {
 		let mut vec = Vec::new();
 		vec.resize(16 * 1024, 0xAA);
 		Some(vec)
 	}
 
-	fn return_option_bytes() -> Option<bytes::Bytes> {
+	fn return_option_bytes() -> AllocateAndReturnByCodec<Option<bytes::Bytes>> {
 		let mut vec = Vec::new();
 		vec.resize(16 * 1024, 0xAA);
 		Some(vec.into())
 	}
 
 	/// Set the storage at key with value.
-	fn set_storage(&mut self, key: &[u8], data: &[u8]) {
+	fn set_storage(
+		&mut self,
+		key: PassFatPointerAndRead<&[u8]>,
+		data: PassFatPointerAndRead<&[u8]>,
+	) {
 		self.place_storage(key.to_vec(), Some(data.to_vec()));
 	}
 
 	/// Copy `hello` into the given mutable reference
-	fn return_value_into_mutable_reference(&self, data: &mut [u8]) {
+	fn return_value_into_mutable_reference(&self, data: PassFatPointerAndReadWrite<&mut [u8]>) {
 		let res = "hello";
 		data[..res.as_bytes().len()].copy_from_slice(res.as_bytes());
 	}
 
 	/// Returns the input data wrapped in an `Option` as result.
-	fn return_option_input(data: Vec<u8>) -> Option<Vec<u8>> {
+	fn return_option_input(
+		data: PassFatPointerAndRead<Vec<u8>>,
+	) -> AllocateAndReturnByCodec<Option<Vec<u8>>> {
 		Some(data)
 	}
 
 	/// Get an array as input and returns a subset of this array.
-	fn get_and_return_array(data: [u8; 34]) -> [u8; 16] {
+	fn get_and_return_array(
+		data: PassPointerAndReadCopy<[u8; 34], 34>,
+	) -> AllocateAndReturnPointer<[u8; 16], 16> {
 		let mut res = [0u8; 16];
 		res.copy_from_slice(&data[..16]);
 		res
 	}
 
 	/// Take and fill mutable array.
-	fn array_as_mutable_reference(data: &mut [u8; 16]) {
+	fn array_as_mutable_reference(data: PassPointerAndWrite<&mut [u8; 16], 16>) {
 		data.copy_from_slice(&TEST_ARRAY);
 	}
 
 	/// Returns the given public key as result.
-	fn return_input_public_key(key: Public) -> Public {
+	fn return_input_public_key(
+		key: PassPointerAndReadCopy<Public, 32>,
+	) -> AllocateAndReturnPointer<Public, 32> {
 		key
 	}
 
 	/// A function that is called with invalid utf8 data from the runtime.
 	///
 	/// This also checks that we accept `_` (wild card) argument names.
-	fn invalid_utf8_data(_: &str) {}
+	fn invalid_utf8_data(_: PassFatPointerAndRead<&str>) {}
 
 	/// Overwrite the native implementation in wasm. The native implementation always returns
 	/// `false` and the replacement function will return always `true`.
 	fn overwrite_native_function_implementation() -> bool {
 		false
 	}
 
-	/// Gets an `u128` and returns this value
```

Comment:
nit: maybe it is worth keeping some test on how to marshal 128 bit integers in backward compatible way - just for reference purposes?

Reply:
Hmm... well, I got rid of this because marshaling `u128`s is not supported anymore, and it's not supported because a) it used the allocator (which was totally silly, as it's a small, constant length type), and b) nothing used it anyway**. (:

If anyone complains I'll add it back in a way that doesn't use the allocator.

** - besides one host function in the benchmarking machinery, which I modified in this PR.